addons/promote/common/src/main/java/org/commonjava/indy/promote/data/PromotionHelper.java

```diff
@@ -0,0 +1,266 @@
+import org.commonjava.indy.content.ContentManager;
+import org.commonjava.indy.content.DownloadManager;
+import org.commonjava.indy.data.IndyDataException;
+import org.commonjava.indy.data.StoreDataManager;
+import org.commonjava.indy.model.core.ArtifactStore;
+import org.commonjava.indy.model.core.Group;
+import org.commonjava.indy.model.core.StoreKey;
+import org.commonjava.indy.model.galley.KeyedLocation;
+import org.commonjava.indy.promote.metrics.PathGauges;
+import org.commonjava.indy.promote.model.PathsPromoteRequest;
+import org.commonjava.indy.promote.model.PathsPromoteResult;
+import org.commonjava.indy.promote.validate.PromotionValidationException;
+import org.commonjava.indy.util.LocationUtils;
+import org.commonjava.maven.galley.event.EventMetadata;
+import org.commonjava.maven.galley.model.ConcreteResource;
+import org.commonjava.maven.galley.model.Transfer;
+import org.commonjava.maven.galley.spi.nfc.NotFoundCache;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+@ApplicationScoped
+public class PromotionHelper
+{
+    private final Logger logger = LoggerFactory.getLogger( getClass() );
+
+    public PromotionHelper()
+    {
+    }
+
+    @Inject
+    private StoreDataManager storeManager;
+
+    @Inject
+    private NotFoundCache nfc;
+
+    @Inject
+    private PathGauges pathGauges;
+
+    @Inject
+    private DownloadManager downloadManager;
+
+    @Inject
+    private ContentManager contentManager;
+
+    public PromotionHelper( StoreDataManager storeManager, DownloadManager downloadManager,
+                            ContentManager contentManager, NotFoundCache nfc )
+    {
+        this.storeManager = storeManager;
+        this.downloadManager = downloadManager;
+        this.contentManager = contentManager;
+        this.nfc = nfc;
+        this.pathGauges = new PathGauges();
+    }
+
+    /**
+     * Clear NFC for the source store and affected groups.
+     *
+     * @param sourcePaths The set of paths that need to be cleared from the NFC.
+     * @param store The store whose affected groups should have their NFC entries cleared
+     */
+    public void clearStoreNFC( final Set<String> sourcePaths, ArtifactStore store )
+    {
+        Set<String> paths = sourcePaths.stream()
+                                       .map( sp -> sp.startsWith( "/" ) && sp.length() > 1 ? sp.substring( 1 ) : sp )
+                                       .collect( Collectors.toSet() );
+
+        paths.forEach( path -> {
+            ConcreteResource resource = new ConcreteResource( LocationUtils.toLocation( store ), path );
+            logger.trace( "Clearing NFC path: {} from: {}\n\tResource: {}", path, store.getKey(), resource );
+            nfc.clearMissing( resource );
```

Comment:
We should consider it definitely. But I did not see any code here in promotion manager to handle that. I thought it might be done via file events? I will keep this open and have a close look at it.

Reply:
Yes, I think file events are handling this part. I'll consider this resolved.